orca/orca-spec.md

```diff
@@ -0,0 +1,278 @@
+
+It could also be specified in the `endpoint-load-metrics-bin` header without BIN prefix:
+
+`endpoint-load-metrics-bin: CZqZmZmZmbk/MQAAAAAAAABAQg4KA2ZvbxGamZmZmZm5P0IOCgNiYXIRmpmZmZmZyT8=`
+
+### Native HTTP
+
+Comma separated key-value pairs in `endpoint-load-metrics`. This is a flattened
+representation of OrcaLoadReport, with the map fields elided into the top level scope by prepending
+the ‘<map_name>.’:
+
+`endpoint-load-metrics: TEXT cpu_utilization=0.3, mem_utilization=0.8, rps_fractional=10.0, eps=1,
+named_metrics.custom_metric_util=0.4`
+
+### JSON
+
+The JSON encoding of OrcaLoadReport:
+
+`endpoint-load-metrics: JSON {“cpu_utilization”: 0.3, “mem_utilization”: 0.8, “rps_fractional”: 10.0,
+“eps”: 1, “named_metrics”: {“custom-metric-util”: 0.4}}`
+
+## Out-of-Band (OOB)
+
+OOB reporting involves an additional load reporting agent that does not sit in the request path.
+This is periodically sampled with sufficient frequency to provide temporal association with
+requests. The data plane LB may either initiate a connection to the agent or offer a gRPC service
+for agents to report to:
+
+```
+// Out-of-band (OOB) load reporting service for the additional load reporting
+// agent that does not sit in the request path. Reports are periodically sampled
+// with sufficient frequency to provide temporal association with requests.
+// OOB reporting compensates the limitation of in-band reporting in revealing
+// costs for backends that do not provide a steady stream of telemetry such as
+// long running stream operations and zero QPS services. This is a server
+// streaming service, client needs to terminate current RPC and initiate
+// a new call to change backend reporting frequency.
+service OpenRcaService {
+  rpc StreamCoreMetrics(OrcaLoadReportRequest) returns (stream xds.data.orca.v3.OrcaLoadReport);
+}
+
+message OrcaLoadReportRequest {
+  // Interval for generating Open RCA core metric responses.
+  google.protobuf.Duration report_interval = 1;
+  // Request costs to collect. If this is empty, all known requests costs tracked by
+  // the load reporting agent will be returned. This provides an opportunity for
+  // the client to selectively obtain a subset of tracked costs.
+  repeated string request_cost_names = 2;
+}
+```
+
+The definition for this service can be found at
+https://github.com/cncf/xds/blob/main/xds/service/orca/v3/orca.proto.
+
+Alternatively, an HTTP endpoint can be exposed by the load reporting server. A GET or HEAD request
+to this endpoint will return a response with the `endpoint-load-metrics` header that contains the
+load report in one of the formats listed above.
+
+# ORCA Integration Considerations
+
+## Sidecar
+
+A sidecar, e.g. Envoy or some other in-container proxy process, can inject the inline ORCA header
+at response completion.
+
+![Sidecar Integration Diagram](./orca-spec-sidecar.jpg)
+
+The advantage of a sidecar model is that the application requires no modification as in the example
```

Comment:
I think this is true only if the sidecar is capable of determining the metrics that the endpoint wishes to report to its clients.  For example, this would probably work for things like CPU or memory utilization, which the sidecar can determine in a fairly generic way.  But it won't work for application-specific things, like custom utilization metrics or query cost metrics -- those will always require cooperation on the part of the application.